chain/store/newName.go

```diff
@@ -0,0 +1,261 @@
+package store
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/nknorg/nkn/common/serialization"
+)
+
+type nameInfo struct {
+	registrant []byte
+	expiresAt  uint32
+}
+
+type namesCleanup map[string]struct{}
+
+func (ni *nameInfo) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, ni.registrant); err != nil {
+		return err
+	}
+	if err := serialization.WriteUint32(w, ni.expiresAt); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ni *nameInfo) Deserialize(r io.Reader) error {
+	var err error
+	ni.registrant, err = serialization.ReadVarBytes(r)
+	if err != nil {
+		return err
+	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ni *nameInfo) Bytes() ([]byte, error) {
+	buff := bytes.NewBuffer(nil)
+	err := ni.Serialize(buff)
+	if err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
+}
+
+func (ni *nameInfo) String() string {
+	return fmt.Sprintf("%064x %d", ni.registrant, ni.expiresAt)
+}
+
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
+
+func getNewNameId(name string) []byte {
+	nameId := sha256.Sum256([]byte(strings.ToLower(name)))
+	return nameId[:hashPrefixLength]
+}
+
+func (sdb *StateDB) getRegistrant(name string) ([]byte, uint32, error) {
+	nameId := getNewNameId(name)
+	ni, err := sdb.getNameInfo(nameId)
+	if err != nil {
+		return nil, 0, errors.New("name registration info not found")
+	}
+
+	return ni.registrant, ni.expiresAt, nil
+}
+
+func (cs *ChainStore) GetRegistrant(name string) ([]byte, uint32, error) {
+	return cs.States.getRegistrant(name)
+}
+
+func (sdb *StateDB) getNameInfo(nameId []byte) (*nameInfo, error) {
+	if v, ok := sdb.names.Load(string(nameId)); ok {
+		if ni, ok := v.(*nameInfo); ok {
+			return ni, nil
+		}
+	}
+
+	enc, err := sdb.trie.TryGet(append(NamePrefix, nameId...))
+	if err != nil {
+		return nil, err
+	}
+	if len(enc) > 0 {
+		ni := new(nameInfo)
+		buff := bytes.NewBuffer(enc)
+		if err := ni.Deserialize(buff); err != nil {
+			return nil, err
+		}
+		sdb.names.Store(string(nameId), ni)
+		return ni, nil
+	}
+
+	return &nameInfo{}, nil
+}
+
+func (sdb *StateDB) registerName(name string, registrant []byte, expiresAt uint32) error {
+	nameId := getNewNameId(name)
+
+	ni, err := sdb.getNameInfo(nameId)
+	if err != nil {
+		return err
+	}
+
+	if !ni.Empty() {
+		return errors.New("name is already registered")
+	}
+	ni.registrant = registrant
+	ni.expiresAt = expiresAt
+
+	sdb.names.Store(string(nameId), ni)
+
+	return sdb.cleanupNamesAtHeight(expiresAt, nameId)
+}
+
+func (sdb *StateDB) updateNameInfo(nameId string, ni *nameInfo) error {
+	nibytes, err := ni.Bytes()
+	if err != nil {
+		return err
+	}
+	return sdb.trie.TryUpdate(append(NamePrefix, nameId...), nibytes)
```

Comment:
Should we update `sdb.names.Store` also after `sdb.trie.TryUpdate` succeed?
Otherwise sub.getNameInfo(id) maybe will return **outdated** nameInfo from below case:
```

 func (sdb *StateDB) getNameInfo(nameId []byte) (*nameInfo, error) {
 	if v, ok := sdb.names.Load(string(nameId)); ok {
 		if ni, ok := v.(*nameInfo); ok {
 			return ni, nil  // <== Here
 		}
 	}
```

Reply:
`sdb.names.Store` will be cleaned up while statedb `FinalizeNames` which will call `TryUpdate` to write name info to the disk.So this scencrio wont exists